src/vstart.sh

```diff
@@ -1697,8 +1715,15 @@ if [ "$ceph_osd" == "crimson-osd" ]; then
         extra_seastar_args=" --trace"
     fi
     if [ "$(expr $(nproc) - 1)" -gt "$(($CEPH_NUM_OSD * crimson_smp))" ]; then
-      echo "crimson_alien_thread_cpu_cores:" $(($CEPH_NUM_OSD * crimson_smp))-"$(expr $(nproc) - 1)"
-      $CEPH_BIN/ceph -c $conf_fn config set osd crimson_alien_thread_cpu_cores $(($CEPH_NUM_OSD * crimson_smp))-"$(expr $(nproc) - 1)"
+        if [ $crimson_alien_num_cores -gt 0 ]; then
+            alien_bottom_cpu=$(($CEPH_NUM_OSD * crimson_smp))
+            alien_top_cpu=$(( alien_bottom_cpu + crimson_alien_num_cores - 1 ))
+            echo "crimson_alien_thread_cpu_cores: $alien_bottom_cpu-$alien_top_cpu"
+            $CEPH_BIN/ceph -c $conf_fn config set osd crimson_alien_thread_cpu_cores "$alien_bottom_cpu-$alien_top_cpu"
```

Comment:
If there are 2 CPU sockets, each with 16 physical cores and are hyper-threaded, the range might be:
CPU0: 0-15, 32-47
CPU1: 16-31,48-63

This means the range of logical cpu cores from the physical ones may not be consecutive. If want to pin the seastar threads to 8 physical cores (CPU0 0-7), and pin alien threads to the rest of 8 physical cores on CPU0, the proper way might need to set them to 8-15,40-47.

What do you think?

Reply:
I think that if we need a precise cpu allocation we can manually set `crimson_alien_thread_cpu_cores` (and override the "primitive" ones already set here by vstart).
The new `crimson_alien_num_cores` vstart flag is introduced only to enable limiting of cores used IIUC.

Reply:
> The new crimson_alien_num_cores vstart flag is introduced only to enable limiting of cores used IIUC.

I agree, and it might be a step further to improve this limitation based on physical cores rather than the logical ones.

Reply:
I see, I was wondering about that, whether the alien threads required to be pinned to the "closer" (that is, in the same socket preferably) as the I/O reactor/seastar threads.
What would the general rule be? In principle, pin always on physical cores, in the same socket (if available) to avoid excessive memory bandwidth between sockets. Anything else?

Reply:
> Would be preferable if given a seastar thread pinned to a physical core X, its associated alien(s) thread to be pinned to the hyperthreaded sibling of core X?

My instinct is that alien threads (blockable) should not share the same physical cores with seastar threads (nonblocking). Seastart threads only do nonblock executions, but hyperthreading won't duplicate the main execution resources. I think this is also why seastar threads uses the one-thread-per-core model. Other than that, closer should be better.

Reply:
Here is a comparison of the effect of modifying the number of alien threads, and number of cores to alien threads (from the example shown in the Description):

```bash
# MDS=0 MON=1 OSD=1 MGR=1 ../src/vstart.sh --new -x --localhost --without-dashboard -o timeout=10000 -o session_timeout=10000 --bluestore --crimson --redirect-output --bluestore-devs /dev/sdc --crimson-smp 4 --crimson-alien-num-threads 2 --crimson-alien-num-cores 1
```

The table on the left shows the reference baseline configuration results; whereas the table on the right hand side, presents the corresponding results on the new configuration with a single core for alien threads. 
Note the values  `(iodepth x num_jobs)`. the Multiple FIO case used 8 processes, each exercising its own RBD image.
![Screenshot 2024-05-20 at 12 24 11](https://github.com/ceph/ceph/assets/23522684/528ef187-44a7-4134-b1dc-adb8e3e9e692)


Is interesting to note the impact on performance of the new configuration. For example, the following takes a closer look to random read 4k, where the new configuration impacts approximately 25% in throughput: (each row represents the output of a single FIO process) 

![Screenshot 2024-05-20 at 12 24 47](https://github.com/ceph/ceph/assets/23522684/836c82c5-1d4d-494c-8608-0dbf76ffdf1f)

Reply:
> I see, so the current logic should change: currently, we first allocate and pin each I/O reactor to its CPU core, then we allocate the alien threads (they can use any available CPU core left). But now, we need to allocate and pin each sibling alien threads for each I/O reactor (aka seastar) thread, on physical cores only.

In a system of 2 CPU sockets, 8 cores per socket, 2 hyperthreaded logical cores per physical core, the CPU range from OS perspective might be:
* CPU0: 0-7,16-23
* CPU1: 8-15,24-31

As for p2l mapping, e.g. the physical core p0 is mapped to logical [0,16], p1 to l[1,17], ..., p8 to [8,24], etc. (refer to `/proc/cpuinfo`).

If there are 2 OSDs, `--crimson-smp`==6 (each OSD has 6 I/O reactor threads, totally 12), alien threads is set to 4 per OSD (totally 8).

If I understand correctly, the current logic will allocate logical cores [0-5] to OSD0 reactors, [6-11] to OSD1 reactors, and the rest of logical cores [12-31] (probably with a limit from `crimson_alien_num_cores`) to the 8 alien threads.

I think the issue of the current logical-core-based arrangement is the range of alien threads [12-31] means they are allowed to run on the physical cores already running the reactor threads, i.e. [16-27] is shared by p0-11.

Towards physical-core-based arrangement, the only change here might be setting the rest of **physical** cores to the 8 alien threads, which are the logical ones [12-23,28-31] (Alien threads are blockable, so they should be able to benefit from hyperthreading).

As @Matan-B suggested, it is also possible to manually set `crimson_alien_thread_cpu_cores` for the physical-core-based cpu allocation. Also, the impact from the inaccurate policy might be negligible in reality. 

> Other than that, closer should be better.

I think this is actually challenging for the best strategy because alien threads also talk to each other and the storage device, and reactor threads also talk to each other and the network device.

Reply:
I think the current PR looks good enough, and as @Matan-B suggested, it is possible to configure `-o "crimson_alien_thread_cpu_cores=xxx"` to experiment advanced core allocations (even with multiple OSDs).

(So the below discussion may not be related to this PR)

Sorry I'm not able to follow the above graph. The typical arrangement of (2 CPU sockets, 8 physical cores per socket, 2 hyperthreads per physical core) should look like below (from `lscpu`):
```
...
NUMA:
  NUMA node(s):          2
  NUMA node0 CPU(s):     0-7,16-23
  NUMA node1 CPU(s):     8-15,24-31
...
```
Node0 means socket0, and node1 means socket1.

And I'm not sure why the policy on the left (socket0) can be different from right (socket1) at the same time, while it looks that all the threads come from the same OSD with 8 reactor-threads and 8 alien-threads.

> for each socket, let n be the number of physical cores per socket (assume that each socket has the same n number of hyper-threading cores)

I think it is incorrect to assume that there are n physical cores and n extra hyper-threading cores. From OS perspective, if HT is enabled, all the vCPUs are hyperthreaded: in the above example, vCPU0 and vCPU16 may be symmetric and are both associated to the same physical core in socket0.

Or probably I'm missing something?

Reply:
Sorry for my abuse of the notation (which caused the confusion), what I meant was the following: I think that the vstart.sh script uses the _processor id_ to identify a cpu to pin the corresponding thread t. The processor id is the second column from the `/proc/cpuinfo `output, I abused the notation to differentiate between the _physical_ and the _hyperthreaded_, from the lscpu output, I called cpus 0 to 7 as physical, and cpus 16-23 hyperthreaded, so _n_ in this case is 8 (because we have the same number of physical and hyperthreading).

```bash
core id  : 0 processor : 0 physical id : 0
:
core id  : 12 processor : 7 physical id : 0
core id  : 0 processor : 8 physical id : 1
:
core id  : 12 processor : 15 physical id : 1
core id  : 0 processor : 16 physical id : 0 <- HT sibling of processor 0
:
core id  : 12 processor : 23 physical id : 0
core id  : 0 processor : 24 physical id : 1 <- HT sibling of processor 8
:
```

In other words, I was suggesting that the script only use the list of _processor id_ [ 0-15], but do not use the ids [16-31]. That was the meaning in the tables above. I was only suggesting two approaches:

-  to allocate an alien thread "close" to its seastar/reactor partner (assuming a 1-1 correspondence, which might not be the optimal case), or 

- to pin first all the seastar/reactors to the available "physical" [0-15], and then to pin the alien threads to the remaining "physical" processor ids -- again, from the list [0-15], if any left.  I guess in the case of 16 OSD, we need to pin the alien threads to HT processor ids because none of the [0-15] have been left -- unless we set that the max number of seastar/reactor threads for the above [0-15] case is 8.

I hope this makes sense 😃

Reply:
Thanks for clarifying!

> * to allocate an alien thread "close" to its seastar/reactor partner (assuming a 1-1 correspondence, which might not be the optimal case), or

It doesn't seem reasonable to me: for 1 OSD with m reactor threads and n alienstore threads, I think a reactor thread is possible talk to all the alienstore threads, and the n alienstore threads may share resources with each other.

> * to pin first all the seastar/reactors to the available "physical" [0-15], and then to pin the alien threads to the remaining "physical" processor ids -- again, from the list [0-15], if any left. I guess in the case of 16 OSD, we need to pin the alien threads to HT processor ids because none of the [0-15] have been left -- unless we set that the max number of seastar/reactor threads for the above [0-15] case is 8.

Yeah, sharing the physical cores may be inevitable in some cases -- in case of 16 physical cores for 16 crimson OSDs, probably the better way is to use a better CPU(s) if the workload is CPU-bounded.